### PR TITLE
Add option to use containerized app with Docker

### LINKED ISSUE #44: Support Docker for isolation, portability, and quick setup
Opening this (and working on it) for the following reasons:

1. Isolation - no need to pollute your global with different versions of "stuff"
2. Portability - reduce "works on my machine" weirdness across environments
3. Quick setup - Minimize install requirements to quickly spin up an example

### PR DESCRIPTION
**Pull request by mtheoryx**
There's a couple of things going on with this PR, so I want to lay it out for anyone unfamiliar with Docker.

**Why?**
- Isolation. You don't even need to install node on your machine.
- Dependency control. Like package.json, even the base OS and node version are specified in code.
- Environment parity and portability. You will get the exact same running container no matter what OS you or your contributor are using.

The Dockerfile, think of it like ingredients for making cupcakes!

This is a line-by-line set of instructions to build an application image. This image can be versioned and pushed to a Registry, allowing others to use it, or even build on top of it.

The docker-compose.yml file. This is your instructions on when to add ingredients, and how to bake and ice the cupcakes!

**A few things to note!**

Because containers are stateless, we are mounting the application code from your machine (the host) into the container. When things change in the container... they are persisted on your host. 

The SIMULATING mode is still valid! You can either use the .env file, or you can pass in an override to turn simulating mode on. I updated the readme for this, too, but it looks like:

```bash
SIMULATING=true docker-compose up --build
```

Please feel free to try the new instructions in the README, and let me know how it works for you! And ask any questions right here in the PR, I'd be happy to explain/answer them.

This should close #44 and should address the node version concern in #40

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+node_modules
```
Comment (mtheoryx) [inline .dockerignore]: This just saves space in the docker container as Docker does not care about these files/directories.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  app:
+    build: .
+    image: electic-io
+    container_name: electric-io
+    command: npm run start
+    ports:
+      - 3000:3000
+    env_file:
+      - .env
+    environment:
```
Comment (mtheoryx) [inline docker-compose.yml]: This allows the environment override for simulating. I could expand this for any of the .env options if we need to. I figure people can modify their own .env files for the other settings and not check them in if they are contributing, but toggling simulating seems like something you'd want to do more often.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      - SIMULATING
+    volumes: # For file editing, watching, etc
```
Comment (mtheoryx) [inline docker-compose.yml]: This preserves the changes made in the container back to your host. And also changes (as you're developing) on your host get reflected back into the running app container.

File: package.json [hunk 4]
```diff
@@ -20,8 +20,9 @@
 		"socket.io": "*"
 	},
 	"engines": {
-		"node": "8"
+		"node": ">= 6"
```
Comment (mtheoryx) [inline package.json]: Per discussion in #40 we settled on a minimum version of 6. I do know that 8.x works as well, so went with >=.

**Comment by noopkat** [comment]
@mtheoryx this is an amazing first pull request - thanks so much! I'm excited about the potential for this project now being container compatible. Let's chat about putting this on Docker Hub asap 💯 

bitmoji

**Pull request merged**